zkevm-circuits/src/evm_circuit/op_execution/arithmetic/bitwise.rs

```diff
@@ -0,0 +1,270 @@
+use super::super::super::FixedLookup;
+use super::super::{Case, Cell, Constraint, ExecutionStep, Word};
+use super::super::utils::common_cases::{OutOfGasCase, StackUnderflowCase};
+use super::super::utils::constraint_builder::ConstraintBuilder;
+use super::super::utils::StateTransition;
+use super::super::{
+    utils, CaseAllocation, CaseConfig, CoreStateInstance, OpExecutionState, OpGadget,
+};
+use crate::impl_op_gadget;
+use crate::util::{Expr, ToWord};
+use bus_mapping::evm::{GasCost, OpcodeId};
+use halo2::plonk::Error;
+use halo2::{arithmetic::FieldExt, circuit::Region};
+use std::convert::TryInto;
+
+static STATE_TRANSITION: StateTransition = StateTransition {
+    gc_delta: Some(3), // 2 stack pops + 1 stack push
+    pc_delta: Some(1),
+    sp_delta: Some(1),
+    gas_delta: Some(GasCost::FASTEST.as_usize()),
+};
+const NUM_POPPED: usize = 2;
+
+#[derive(Clone, Debug)]
+struct BitwiseSuccessCase<F> {
+    case_selector: Cell<F>,
+    a: Word<F>,
+    b: Word<F>,
+    c: Word<F>,
+}
+
+macro_rules! impl_bitwise_op_gadget {
+    ($op:ident) => {
+        paste::paste! {
+            impl_op_gadget!(
+                #set[ [<$op:upper>] ]
+                [<$op Gadget>] {
+                    [<$op SuccessCase>](),
+                    StackUnderflowCase(NUM_POPPED),
+                    OutOfGasCase(STATE_TRANSITION.gas_delta.unwrap()),
+                }
+            );
+
+            #[derive(Clone, Debug)]
+            struct [<$op SuccessCase>]<F> {
+                case_selector: Cell<F>,
+                a: Word<F>,
+                b: Word<F>,
+                c: Word<F>,
+            }
+
+            impl<F: FieldExt> [<$op SuccessCase>]<F> {
+                pub(crate) const CASE_CONFIG: &'static CaseConfig = &CaseConfig {
+                    case: Case::Success,
+                    num_word: 3, // a + b + c
+                    num_cell: 0,
+                    will_halt: false,
+                };
+
+                pub(crate) fn construct(alloc: &mut CaseAllocation<F>) -> Self {
+                    Self {
+                        case_selector: alloc.selector.clone(),
+                        a: alloc.words.pop().unwrap(),
+                        b: alloc.words.pop().unwrap(),
+                        c: alloc.words.pop().unwrap(),
+                    }
+                }
+
+                pub(crate) fn constraint(
+                    &self,
+                    state_curr: &OpExecutionState<F>,
+                    state_next: &OpExecutionState<F>,
+                    name: &'static str,
+                ) -> Constraint<F> {
+                    let mut cb = ConstraintBuilder::default();
+
+                    // for idx in 0..32 {
+                    //     cb.add_fixed_lookup(
+                    //         FixedLookup::[<Bitwise $op>],
```

Comment:
I don't know if this will matter, but if the only difference between these bitwise opcodes is the lookup table, perhaps it's a bit more efficient to just change the `tag` of `FixedLookup` to an expression and set the value to the correct table depending on the opcode being handled?

Reply:
`FixedLookup` also includes range lookup tables. So we cannot use the opcode as the tag unless we separate them from the lookup tables. But I think it's a bit overkill.

Reply:
The only difference is the `tag` value for all fixed lookups no? Which is unique per table so I think it should work without splitting them up. The tag would be easy and cheap to calculate as `opcode.as_u8() - OpcodeId::AND.as_u8() + BitwiseAnd as u8`.

If this is overkill or not I don't know, I have no idea what kinds of overheads additional opcode implementations bring, but this seems like a minor change to be able to reuse all constraints between those opcodes.